HW2/P5/physics.pyx

```diff
@@ -90,40 +136,68 @@ cpdef update(FLOAT[:, ::1] XY,
              float t):
     cdef:
         int count = XY.shape[0]
-        int i, j, dim
+        int i, j, dim,xgrid,ygrid,grid_size
         FLOAT *XY1, *XY2, *V1, *V2
         # SUBPROBLEM 4: uncomment this code.
-        # omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
+        omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
 
     assert XY.shape[0] == V.shape[0]
     assert XY.shape[1] == V.shape[1] == 2
 
-    with nogil:
-        # bounce off of walls
-        #
-        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
-        # scheduling.
-        for i in range(count):
-            for dim in range(2):
-                if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
-                    ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
-                    V[i, dim] *= -1
-
-        # bounce off of each other
-        #
-        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
-        # scheduling.
-        for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
-
-        # update positions
-        #
-        # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
-        #    scheduling).
-        # SUBPROBLEM 2: update the grid values.
-        for i in range(count):
-            for dim in range(2):
-                XY[i, dim] += V[i, dim] * t
+
+    # bounce off of walls
+    #
+    # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+    # scheduling.
+    #for i in range(count):
+    for i in prange(count, nogil=True, schedule='static', chunksize = count/4, num_threads = 4):
+        for dim in range(2):
+            if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
+                ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
+                V[i, dim] *= -1
+
+    # bounce off of each other
+    #
+    # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+    # scheduling.
+    #for i in range(count):
+    for i in prange(count,nogil=True, schedule='static', chunksize = count/4, num_threads = 4):
+        sub_update(XY, V, R, i, count, Grid, grid_spacing,locks)
+
+    # update positions
+    #
+    # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
+    #    scheduling).
+    # SUBPROBLEM 2: update the grid values.
+    #for i in range(count):
+
+    #so at first, all grids should be set to -1
+    #then update with the newest location
+    Grid[:] = -1
```

Comment:
I've seen several solutions to this, but this one is definitely the most concise. Well done!